src/tscache.ts

```diff
@@ -78,16 +78,16 @@ export class TsCache
 	private syntacticDiagnosticsCache!: ICache<IDiagnostics[]>;
 	private hashOptions = { algorithm: "sha1", ignoreUnknown: false };
 
-	constructor(private noCache: boolean, hashIgnoreUnknown: boolean, private host: tsTypes.LanguageServiceHost, private cacheRoot: string, private options: tsTypes.CompilerOptions, private rollupConfig: any, rootFilenames: string[], private context: RollupContext)
+	constructor(private noCache: boolean, runClean: boolean, hashIgnoreUnknown: boolean, private host: tsTypes.LanguageServiceHost, private cacheRoot: string, private options: tsTypes.CompilerOptions, private rollupConfig: any, rootFilenames: string[], private context: RollupContext)
 	{
 		this.dependencyTree = new Graph({ directed: true });
 		this.dependencyTree.setDefaultNodeLabel((_node: string) => ({ dirty: false }));
 
-		if (noCache)
-		{
+		if (runClean)
```

Comment:
Does skipping `this.clean()` have a significant performance improvement?
I imagine that after the first clean, the performance hit would be negligible

Reply:
Yeah, this is more to avoid surprising behavior -- cleaning old caches and disabling caching altogether are different actions, but they were joined together before

Reply:
> cleaning old caches and disabling caching altogether are different actions

Yea `clean` is a bit of an edge-case as it's a side-effect. Ideally one would `clean` in a totally separate command/export, but there's no conventional way of doing that in Rollup. So while a bit of a head-scratcher initially, it did eventually make sense to me. When I was optimizing the cache, I thought about changing this for a bit but ended up leaving it as is.

> to avoid surprising behavior

To an experienced user, I thought it might be surprising that something runs without caching and _doesn't_ clean. Since they have been joined at the hip for so long.

But this turns off caching _implicitly_ with no config, so I suppose we can ignore that scenario.
If `watch` removed caches when it previously didn't, that would probably be _more_ surprising.

Had to write the logic out there a bit, but that makes sense to me now. Removing a head-scratcher (partially) required some head-scratching 😅

<hr>

We could potentially expose this as an option to completely eliminate head-scratching so that `cleanCache` and `noCache` are entirely separate config options 🤔 It would technically be a breaking change though to fully separate those.

When I was reviewing this, I remembered there was some reason I never exposed it myself. I think it was potential bugginess of the dependency tree with no cache, but I think I might have fixed all those code paths in #362 et al... Testing my memory a bit here 😅